builtin/gc.c

```diff
@@ -1345,6 +1345,154 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
 	return 0;
 }
 
+
+static void copy_file(const char *srcdir, const char *dstdir, const char *name)
+{
+	int ret = 0;
+	struct strbuf src = STRBUF_INIT, dst = STRBUF_INIT;
+	int srcfd = -1, dstfd = -1;
+	char buf[1024];
+	ssize_t nr;
+
+	strbuf_addf(&src, "%s/%s", srcdir, name);
+	strbuf_addf(&dst, "%s/%s", dstdir, name);
+
+	srcfd = open(src.buf, O_RDONLY);
+	if (srcfd < 0) {
+		error("failed to open source file");
+		ret = 1;
+		goto cleanup;
+	}
+
+	dstfd = open(dst.buf, O_WRONLY | O_CREAT | O_TRUNC, 0644);
+	if (dstfd < 0) {
+		error("failed to open destination file");
+		ret = 1;
+		goto cleanup;
+	}
+
+	while ((nr = read(srcfd, buf, sizeof(buf))) > 0) {
+		if (write(dstfd, buf, nr) < 0) {
+			error("failed to write to destination file");
+			ret = 1;
+		}
+	}
+
+cleanup:
+	if (srcfd >= 0) close(srcfd);
+	if (dstfd >= 0) close(dstfd);
+
+	if (ret)
+		die_errno(_("failed to copy '%s' to '%s'"), src.buf, dst.buf);
+
+	strbuf_release(&src);
+	strbuf_release(&dst);
+}
+
+static void move_file(const char *srcdir, const char *dstdir, const char *name)
+{
+	struct strbuf src = STRBUF_INIT, dst = STRBUF_INIT;
+
+	strbuf_addf(&src, "%s/%s", srcdir, name);
+	strbuf_addf(&dst, "%s/%s", dstdir, name);
+
+	if (rename(src.buf, dst.buf))
+		die_errno(_("failed to move '%s' to '%s'"), src.buf, dst.buf);
+
+	strbuf_release(&src);
+	strbuf_release(&dst);
+}
+
+static void delete_file(const char *dir, const char *name)
+{
+	struct strbuf path = STRBUF_INIT;
+
+	strbuf_addf(&path, "%s/%s", dir, name);
+
+	if (unlink(path.buf))
+		warning_errno(_("failed to delete '%s'"), path.buf);
+
+	strbuf_release(&path);
+}
+
+static void migrate_pack(const char *srcdir, const char *dstdir,
+			 const char *pack_filename)
+{
+	struct strbuf path = STRBUF_INIT;
+	struct stat st;
+	char *basename, *keep_filename, *rev_filename, *idx_filename;
+	int has_keep, has_rev;
+
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
+	keep_filename = xstrfmt("%s.keep", basename);
+	rev_filename = xstrfmt("%s.rev", basename);
+	idx_filename = xstrfmt("%s.idx", basename);
+
+	strbuf_addf(&path, "%s/%s", srcdir, keep_filename);
+	has_keep = !stat(path.buf, &st);
+	strbuf_reset(&path);
+	strbuf_addf(&path, "%s/%s", srcdir, rev_filename);
+	has_rev = !stat(path.buf, &st);
+	strbuf_release(&path);
+
+	/* Copy all but the index file, which we will *move* atomically */
+	copy_file(srcdir, dstdir, pack_filename);
```

Comment:
I would like to try hard-linking it first, i.e. the following strategy:

1. create a hard-link of the `.pack` file to the new location
2. create hard-links of the corresponding `.keep`, `.rev` and finally `.idx` files, too
3. remove the original `.idx` file, then the `.pack`, `.rev` and `.keep` file, too

Of course, hard-linking can fail, e.g. when the object directories are on different drives. In that instance, we do have to fall back to copying instead (but keeping the operations in the same order).